RandBLAS/trig_skops.hh

```diff
@@ -0,0 +1,493 @@
+        std::vector<int64_t> diag = generateRademacherVector_parallel(cols);
+
+        for(int col=0; col < cols; col++) {
+        if(diag[col] > 0)
+            continue;
+        blas::scal(rows, diag[col], &A[col * rows], 1);
+    }
+    }
+
+    template<typename T, SignedInteger sint_t>
+    void applyDiagonalRademacher(
+                                bool left,
+                                blas::Layout layout,
+                                int64_t rows,
+                                int64_t cols,
+                                T* A,
+                                sint_t* diag
+                                )
+    {
+        if(left && layout == blas::Layout::ColMajor) {
+            for(int col=0; col < cols; col++) {
+                if(diag[col] > 0)
+                    continue;
+                blas::scal(rows, diag[col], &A[col * rows], 1);
+            }
+        }
+        else if(left && layout == blas::Layout::RowMajor) {
+            for(int col=0; col < cols; col++) {
+                if(diag[col] > 0)
+                    continue;
+                blas::scal(rows, diag[col], &A[col], cols);
+            }
+        }
+        else if(!left && layout == blas::Layout::ColMajor) {
+            for(int row = 0; row < rows; row++) {
+                if(diag[row] > 0)
+                    continue;
+                blas::scal(cols, diag[row], &A[row], rows);
+            }
+        }
+        else {
+            for(int row = 0; row < rows; row++) {
+                if(diag[row] > 0)
+                    continue;
+                blas::scal(cols, diag[row], &A[row * cols], 1);
+            }
+        }
+    }
+
+    // `applyDiagonalRademacher`should have a similar API as `sketch_general`
+    // Will be called inside of `lskget`
+    // B <- alpha * diag(Rademacher) * op(A) + beta * B
+    // `alpha`, `beta` aren't needed and shape(A) == shape(B)
+    // lda == ldb
+    template<typename T, SignedInteger sint_t>
+    void applyDiagonalRademacher(
+        blas::Layout layout,
+        blas::Op opA, // taken from `sketch_general/lskget`
+        int64_t n, // everything is the same size as `op(A)`: m-by-n
+        int64_t m,
+        // std::vector<int64_t> &diag,
+        sint_t* diag,
+        const T* A, // The data matrix that won't be modified
+        // int64_t lda,
+        T* B // The destination matrix
+        // int64_t ldb
+    ) {
+        // B <- alpha * diag(Rademacher) * op(A) + beta * B
+        // `alpha`, `beta` aren't needed
+        // && shape(A) == shape(B) && lda == ldb && shape({A | B}) = m-by-n
+        //NOTE: Use `layout` and `opA` for working with RowMajor data
+        int64_t lda = m;
+        int64_t ldb = m;
+
+        //NOTE: Should the copy be made inside of `applyDiagonalRademacher` or
+        // should it be inside of `lskget`?
+        lapack::lacpy(lapack::MatrixType::General, m, n, A, lda, B, ldb);
+
+        applyDiagonalRademacher(m, n, B, diag);
+    }
```

Comment:
What's this for? We discussed that the functions you're working on need to do everything in-place.

Reply:
Left it in by accident, removed!